### ISSUE TITLE
Verify the connection ID to avoid malicious attempts to store credentials for another user 

### ISSUE THREAD

**Issue by bastienbeurier** [NangoHQ/nango#582]
Raised by this community [message](https://nango-community.slack.com/archives/C04ABR352H0/p1682327350087349) (and a couple other ones before). 

We should have a mechanism to verify the connection ID to avoid malicious attempts to store credentials for another user in the application.

**Comment by rguldener** [comment]
Intercom has the same issue and solves this with an HMAC of the user id + a secret. I am sure others have this problem as well.

Intercom docs for inspiration:
https://www.intercom.com/help/en/articles/183-enable-identity-verification-for-web-and-mobile
https://gist.github.com/thewheat/7342c76ade46e7322c3e

**Comment by bastienbeurier** [comment]
Suggestion from Simon Stone in the [community](https://nango-community.slack.com/archives/C04ABR352H0/p1682427877815299?thread_ts=1682327350.087349&cid=C04ABR352H0): 
`If the HMAC was over the provider config key and the connection ID, then we'd also be able to verify that the user was allowed to create credentials for that provider.`

@simon, do you want to explicitly whitelist which provider each user can use?

IMO a user_hash based on secret key is enough. This would prevent any attacker who doesn't have the secret key from performing any authorization at all.

**Comment by sstone1** [comment]
@bastienbeurier I can see some scenarios for us in the future where we would only want to allow connections to certain providers based on the product features our user has signed up for, for example if they hadn't signed up for eSignature we would not want them to be able to connect to DocuSign.

**Comment by bastienbeurier** [comment]
Understood. 

Because this requires more work of having this configuration system, could that be possibly handled on your app? e.g. not showing the integration UI to unauthorized accounts? That way build the minimal scope to prevent attacker spoofing accounts?

Otherwise we can do this in two steps: 
- Level 1 security: use a `user_hash` for triggering an OAuth flow
- Level 2 security: whitelist allowed user/provider combinations

Also happy to evaluate alternatives!

**Comment by sstone1** [comment]
@bastienbeurier I think the single hash can handle all the requirements, as long as the hash is over the combination of provider config key and connection ID, rather than just the connection ID. 

My proposal is that we have some new Nango server configuration:

```bash
NANGO_HMAC_REQUIRED=true
NANGO_HMAC_ALGORITHM=sha256
NANGO_HMAC_KEY=<PICK-A-SECRET-KEY>
```

We extend the Nango front-end API to accept a HMAC digest, which is verified by the Nango server:

```js
nango.auth('<CONFIG-KEY>', '<CONNECTION-ID>', { hmac: '<HMAC-DIGEST>' });
```

Then applications must generate HMAC digests with the same algorithm, key, provider config key and connection ID:

```js
const { createHash } = await import('node:crypto');

const hmac = crypto.createHmac('sha256', '<PICK-A-SECRET-KEY>');
hmac.update('<CONFIG-KEY>');
hmac.update('<CONNECTION-ID>');
const digest = hmac.digest('hex');
```

The application should perform whatever authentication and authorization checks it wants to perform before issuing the HMAC digest, and obviously should never reveal the secret key shared with the Nango server.

**Comment by bastienbeurier** [comment]
Resolved by #591, huge kudos to @sstone1 🙌
